miner/Miner.hs

```diff
@@ -208,107 +190,114 @@ main = do
+        u = coordinator a
 
 -- | A supervisor thread that listens for new work and manages mining threads.
 --
-mining :: (TargetBytes -> HeaderBytes -> IO HeaderBytes) -> Env -> WorkBytes -> IO ()
-mining go e wb = do
+mining :: (TargetBytes -> HeaderBytes -> RIO Env HeaderBytes) -> WorkBytes -> RIO Env ()
+mining go wb = do
     race updateSignal (go tbytes hbytes) >>= traverse_ miningSuccess
-    getWork e >>= traverse_ (mining go e)
+    getWork >>= traverse_ (mining go)
   where
-    T3 cbytes tbytes hbytes = unWorkBytes wb
+    T3 (ChainBytes cbs) tbytes hbytes = unWorkBytes wb
 
-    a :: ClientArgs
-    a = args e
+    chain :: IO Utf8Builder
+    chain = display . chainIdInt @Int <$> runGet decodeChainId cbs
 
     -- TODO Rework to use Servant's streaming? Otherwise I can't use the
     -- convenient client function here.
-    updateSignal :: IO ()
-    updateSignal = withEvents req (mgr e) (void . SP.head_ . SP.filter realEvent)
+    updateSignal :: RIO Env ()
+    updateSignal = catchAny f $ \_ -> do
+        logWarn "Couldn't connect to update stream. Trying again..."
+        updateSignal
       where
+        f :: RIO Env ()
+        f = do
+            e <- ask
+            liftIO $ withEvents (req $ args e) (mgr e) (void . SP.head_ . SP.filter realEvent)
+            cid <- liftIO chain
+            logInfo $ cid <> ": Current work was preempted."
+
         -- TODO Formalize the signal content a bit more?
         realEvent :: ServerEvent -> Bool
-        realEvent (ServerEvent _ _ _) = True
+        realEvent ServerEvent{} = True
         realEvent _ = False
 
         -- TODO This is an uncomfortable URL hardcoding.
-        req :: Request
-        req = defaultRequest
-            { host = B.pack . baseUrlHost . _url $ coordinator a
+        req :: ClientArgs -> Request
+        req a = defaultRequest
+            { host = encodeUtf8 . T.pack . baseUrlHost $ coordinator a
             , path = "chainweb/0.0/" <> encodeUtf8 (toText $ version a) <> "/mining/updates"
-            , port = baseUrlPort . _url $ coordinator a
+            , port = baseUrlPort $ coordinator a
             , secure = True
             , method = "GET"
-            , requestBody = RequestBodyBS $ _chainBytes cbytes }
+            , requestBody = RequestBodyBS cbs }
 
     -- | If the `go` call won the `race`, this function yields the result back
     -- to some "mining coordinator" (likely a chainweb-node). If `updateSignal`
     -- won the race instead, then the `go` call is automatically cancelled.
     --
-    miningSuccess :: HeaderBytes -> IO ()
+    miningSuccess :: HeaderBytes -> RIO Env ()
     miningSuccess h = do
-      bh <- bytesToBlockHeader h
-      putStrLn $ "Success!!!  " <> (show $ _blockNonce bh)
-      putStrLn $ "difficulty = " <> (T.unpack $ showTargetHex $ _blockTarget bh)
-      putStrLn $ "blockHash  = " <> (T.unpack $ hashToHex $ _blockHash bh)
-      putStrLn $ "powHash    = " <> (T.unpack $ powHashToHex $ _blockPow bh)
```

Comment:
@mightybyte we can reintroduce these if you like, possibly at the `Debug` level. The day-1 user probably just wants to know if they mined successfully or not, which the `logInfo` call below does.

Reply:
I think day-1 users will probably want to know how to find the block that they mined.  This means at least showing them chain ID and block height.  Also probably block hash.  But then people expect to see zeroes in the block hash, so `powHash` makes sense as well.  Difficulty we could probably leave out.  That was more for my benefit as a sanity check.

Reply:
Very easy to add back as a follow-up.